### PR TITLE
libs: Add `envoy.dependency.check[release-issues/v0.0.1]`

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

### CODE REVIEW

File: envoy.dependency.check/envoy/dependency/check/abstract/issues.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+
+import string
+from functools import cached_property
+
+import abstracts
+
+from aio.api import github
+
+
+GITHUB_REPO_LOCATION = "phlax/envoy"
```
Comment (phlax) [inline envoy.dependency.check/envoy/dependency/check/abstract/issues.py:10]: fixme

**Comment by phlax** [comment]
cc @htuch @moderation

File: envoy.dependency.check/envoy/dependency/check/abstract/issues.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        return await self.issue.comment(comment)
+
+
+class AGithubDependencyIssues(metaclass=abstracts.Abstraction):
+    """Github issues associated with dependencies."""
+
+    def __init__(
+            self,
+            github,
+            body_tpl=BODY_TPL,
```
Comment (phlax) [inline envoy.dependency.check/envoy/dependency/check/abstract/issues.py:143]: add hints

File: envoy.dependency.check/setup.cfg [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+types =
+    mypy
+    mypy-abstracts
+publish = wheel
+
+[options.package_data]
+* = py.typed
+
+[options.entry_points]
+console_scripts =
```
Comment (phlax) [inline envoy.dependency.check/setup.cfg:57]: remove this for now

**Comment by phlax** [comment]
temporarily WIPping this as i have an iteration that includes the cve scanner, and various updates

File: envoy.dependency.check/envoy/dependency/check/abstract/checker.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            lambda d: (
+                d.release.date, ))
+        async for dep in preloader:
+            self.log.debug(f"Preloaded release date: {dep.id}")
+
+    @checker.preload(
+        when=["issues"],
+        catches=[gidgethub.GitHubException])
+    async def preload_issues(self) -> None:
+        # raise Exception("ANOTHER ERROR")
```
Comment (phlax) [inline envoy.dependency.check/envoy/dependency/check/abstract/checker.py:335]: remove

File: envoy.dependency.check/envoy/dependency/check/abstract/dependency.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        """Github version, as parsed from the URL."""
+        if self.url_components[5] != 'archive':
+            # Release tag is a path component.
+            # assert (components[5] == 'releases')
+            return self.url_components[7]
+        # Only support .tar.gz, .zip today. Figure out the release tag
+        # from this filename.
+        if self.url_components[-1].endswith('.tar.gz'):
+            return self.url_components[-1][:-len('.tar.gz')]
+        # assert (components[-1].endswith('.zip'))
```
Comment (phlax) [inline envoy.dependency.check/envoy/dependency/check/abstract/dependency.py:71]: remove this or at least improve the comment, ditto above

File: envoy.dependency.check/envoy/dependency/check/abstract/dependency.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                release=newer_release)
+            if (newer_release
+                and (version.parse(newer_release.tag_name)
+                     != self.release.version))
+            else None)
+
+    @property
+    def organization(self) -> str:
+        """Github organization name."""
+        return self.url_components[3]
```
Comment (kfaseela) [inline envoy.dependency.check/envoy/dependency/check/abstract/dependency.py:110]: should we have all these indices defined as constants so that it is easier to understand?

Comment (phlax) [reply envoy.dependency.check/envoy/dependency/check/abstract/dependency.py:110]: this is one of the bits i least like about this implementation 

what i would like to do is add a `GithubURLParser` that can tell you

- the type of url
- org/user/repo
- additional info according to type

and that would raise appropriate errors

i havent found anything that does this already, but i havent looked that hard either

as this will require a bit of thought - for now i would prefer to keep the existing implementation (im just handling a bit of error handling to it)

if its a blocker tho, we can rethink the url parsing

one consideration here is that the URL is the SoA for the repo connect - this means that configs like here https://github.com/envoyproxy/envoy/pull/19473/files#diff-a2ca502474bb8ec5f092176f7ee6d2f11fc395ebb00e1717d6fbad9b16ee8fd8R481 wont work 

(i will comment this shortly 8/)

tldr i think this needs quite a bit of a rethink - but this is close to what it does now and works to our current expectations, but i would prefer not to formalize it

**Comment by phlax** [comment]
force pushed to fix dco - apologies

**Comment by phlax** [comment]
reWIPping this until #186 lands

**Comment by phlax** [comment]
also waiting for #187

File: envoy.dependency.check/envoy/dependency/check/abstract/issues.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+                yield issue
+
+    @property  # type:ignore
+    @abstracts.interfacemethod
+    def issue_class(self) -> Type[AGithubDependencyIssue]:
+        """Issue class."""
+        raise NotImplementedError
+
+    @async_property(cache=True)
+    async def missing_labels(self) -> Tuple[str, ...]:
```
Comment (htuch) [inline envoy.dependency.check/envoy/dependency/check/abstract/issues.py:198]: There is a lot of logic in this file that I think might also apply to other situations where we want to automagically create, dedupe and label issues. Would it make sense to factor this out?

**Review by htuch — COMMENTED** [review]
LGTM modulo one question.

File: envoy.dependency.check/envoy/dependency/check/abstract/issues.py (hunk 7 above)
Comment (phlax) [reply envoy.dependency.check/envoy/dependency/check/abstract/issues.py:198]: yep, that makes sense - i have made some progress on this already making it more generic so we can cve_issues

the bit im not entirely clear about is where to factor it out to - perhaps to aio.api.github if the logic is generic enough

can i do this as/when i add the cve issues as this interface will require a bit of rethinking then anyway

Comment (htuch) [reply envoy.dependency.check/envoy/dependency/check/abstract/issues.py:198]: Sure, just wanted to flag this as worth considering.

**Review by htuch — APPROVED** [review]

**Review by kfaseela — APPROVED** [review]
